Auto-expand “details” element for find-in-page and scrolling to fragments

nt1m opened the pull request:
#### 23cdbeb3cd776e2337f220c18b86f0ddb639bd50
<pre>
Auto-expand “details” element for find-in-page and scrolling to fragments
<a href="https://bugs.webkit.org/show_bug.cgi?id=228843">https://bugs.webkit.org/show_bug.cgi?id=228843</a>
<a href="https://rdar.apple.com/81856620">rdar://81856620</a>

Reviewed by Darin Adler.

This change causes any closed &lt;details&gt; element to be opened
automatically (auto-expanded) if either:

- the user is using find-in-page and a descendant node of the closed
  &lt;details&gt; element contains a match for the given search string
- the user is navigating to a fragment ID and the fragment ID is for a
  descendant element of the closed &lt;details&gt; element

…per the requirements in the HTML spec’s Find-in-page section at
<a href="https://html.spec.whatwg.org/#interaction-with-details-and-hidden=until-found">https://html.spec.whatwg.org/#interaction-with-details-and-hidden=until-found</a>

Otherwise, without this change, when a descendant node of a closed
&lt;details&gt; element contains a match for the given find-in-page search
string, or when navigating to a fragment ID for a descendant element of
a closed &lt;details&gt; element, the closed &lt;details&gt; element isn’t opened
automatically (not auto-expanded).

Changes to TextIterator ensure that content inside details elements are
not skipped over for the purposes of find-in-page.

This change also introduces the DetailsAutoExpandEnabled preference
(disabled by default), for controlling whether &lt;details&gt; elements get
auto-expanded under the conditions described above.

And this also adds a testRunner.indicateMatchIndex(index) function —
to emulate jumping through matches in the non-findString codepath.

Original PR by: Michael[tm] Smith &lt;mike@w3.org&gt;

* LayoutTests/TestExpectations:
* LayoutTests/editing/find/cocoa/find-and-replace-in-closed-details-expected.txt: Added.
* LayoutTests/editing/find/cocoa/find-and-replace-in-closed-details.html: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-closed-details-expected.txt: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-closed-details.html: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-summary-of-closed-details-expected.txt: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-summary-of-closed-details.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/auto-expand-details-element-fragment-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::TextIterator::TextIterator):
(WebCore::isRendererAccessible):
(WebCore::isConsideredSkippedContent):
* Source/WebCore/editing/TextIteratorBehavior.h:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::revealClosedDetailsAncestors):
* Source/WebCore/html/HTMLDetailsElement.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFragmentInternal):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::autoRevealsWhenFound const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setAutoRevealsWhenFound):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
(WebCore::StyleRareInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageIndicateFindMatch):
* Source/WebKit/UIProcess/API/C/WKPage.h:
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::didFindString):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::decorateTextRangeWithStyle):
* Source/WebKit/WebProcess/WebPage/ios/FindControllerIOS.mm:
(WebKit::FindController::didFindString):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::postPageMessage):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.h:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::indicateFindMatch):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):

Canonical link: <a href="https://commits.webkit.org/296708@main">https://commits.webkit.org/296708@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/79318c2d9002c0ea7604cc5d9a34c4bb45eb4296

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/109380 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/29038 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/19467 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/114585 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/59618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/111343 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/29717 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/37626 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/83128 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/59618 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/112328 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/23660 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/98517 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/63585 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/23044 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/16660 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/59207 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/101879 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/93029 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/16702 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/117697 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/107936 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/36421 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/26957 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/92137 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/36792 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/94779 "1 flakes") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91952 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23412 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/36883 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/14629 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/32226 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/36314 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/41790 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/132202 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/35986 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/35821 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/39324 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/37689 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -453,7 +454,14 @@ static bool isRendererAccessible(const RenderObject* renderer, TextIteratorBehav
+        if (renderer->document().settings().detailsAutoExpandEnabled()) {
+            for (auto& ancestor : ancestorsOfType<RenderElement>(*renderer)) {
+                if (RefPtr element = dynamicDowncast<HTMLElement>(ancestor.element()); is<HTMLDetailsElement>(element))
+                    return true;
```
This looks like a perf hit in the common case where we don't have details in the ancestor chain at all (nit any particular reason for RenderElement? I would think `<details>` needs to be at least a block container) -same below.

darinadler commented on LayoutTests/platform/mac-wk1/TestExpectations:
```diff
@@ -1758,6 +1758,10 @@ http/tests/download/convert-cached-load-to-download.html [ Skip ]
 # Can't load application/octet-stream in WK1
 http/tests/mime/html-with-nosniff-html.html [ Skip ]
 
+# testRunner.findString function is WK2-only, not implemented for WK1.
+editing/text-iterator/find-in-page-in-closed-details.html [ Skip ]
+editing/text-iterator/find-in-page-in-summary-of-closed-details.html [ Skip ]
```
Wouldn’t this belong in `platform/wk1/TestExpectations` instead of `platform/mac-wk1/TestExpectations`?

gsnedders replied:
There is no `platform/wk1` (oddly!), only a `platform/wk2`.

I've kinda wanted to change this for ages, but apparently I've never even filed an issue for this. Here's one now: https://bugs.webkit.org/show_bug.cgi?id=294153.

The best you can do today is put the WebKit1 result in `LayoutTests`, and then the WebKit2 result in `LayoutTests/platform/wk2`.

nt1m replied:
Since other platforms don't have wk1 bots afaik, I think this is probably fine? There's prior art in this file too

darinadler commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -2205,6 +2208,23 @@ Element* Node::enclosingLinkEventParentOrSelf()
+{
+    if (!document().settings().detailsAutoExpandEnabled())
+        return;
+    Vector<RefPtr<HTMLDetailsElement>> detailsElements;
```
```suggestion
    Vector<Ref<HTMLDetailsElement>> detailsElements;
```

darinadler reviewed (commented):
Will this work for the find command that’s implemented in Safari?

darinadler commented on Source/WebCore/dom/Node.h:
```diff
@@ -349,6 +349,8 @@ class Node : public EventTarget, public CanMakeCheckedPtr<Node> {
     // Returns the enclosing event parent Element (or self) that, when clicked, would trigger a navigation.
     WEBCORE_EXPORT Element* enclosingLinkEventParentOrSelf();
 
+    WEBCORE_EXPORT void expandAllDetailsAncestors();
```
This doesn't feel like it should be a `Node` member function to me. Maybe it can be a free function in `HTMLDetailsElement.cpp` or in `FindController.cpp` or perhaps a static member function of one of the classes. It doesn't need any `private` or `protected` access to `Node` and doesn’t seem like a fundamental `Node` operation.

nt1m replied:
I'll move it as a free function in `HTMLDetailsElement`

darinadler commented on Source/WebCore/rendering/style/StyleRareInheritedData.h:
```diff
@@ -151,6 +151,7 @@ class StyleRareInheritedData : public RefCounted<StyleRareInheritedData> {
     unsigned textAlignLast : 3; // TextAlignLast
     unsigned textJustify : 2; // TextJustify
     unsigned textDecorationSkipInk : 2; // TextDecorationSkipInk
+    unsigned mathStyle : 1; // MathStyle
```
What’s the motivation for this change?

nt1m replied:
Just better bit-packing, I could have put my new bit in this extra spot, but my new bit went better with the ones below.

nt1m replied:
I split out this change to https://github.com/WebKit/WebKit/pull/47222

darinadler commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -455,7 +464,10 @@ static bool isRendererAccessible(const RenderObject* renderer, TextIteratorBehav
+        if (style.usedContentVisibility() == ContentVisibility::Auto)
+            return true;
+        if (renderer->document().settings().detailsAutoExpandEnabled() && renderer->style().autoRevealsWhenFound())
+            return true;
```
Why does `autoRevealsWhenFound` need to be precomputed and stored on the RenderStyle? Why not just check         `m_document->settings().detailsAutoExpandEnabled() && is<HTMLDetailsElement>(element)` here, through an appropriate helper function? Similarly, why do the `detailsAutoExpandEnabled()` check twice? Is this solving a performance problem?

nt1m replied:
This is a performance optimization, since the `autoRevealsWhenFound` bit is inherited, this avoids a whole ancestor walk:

```
            for (auto& ancestor : ancestorsOfType<RenderElement>(*renderer)) {
                if (RefPtr element = dynamicDowncast<HTMLElement>(ancestor.element()); is<HTMLDetailsElement>(element))
                    return true;
                }
            }
````

darinadler commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -466,7 +478,13 @@ static bool isConsideredSkippedContent(const RenderBox* renderBox, TextIteratorB
+    if (behaviors.contains(TextIteratorBehavior::EntersSkippedContentRelevantToUser)) {
+        if (renderBox->document().settings().detailsAutoExpandEnabled() && renderBox->style().autoRevealsWhenFound())
+            return false;
+        return renderBox->style().usedContentVisibility() == ContentVisibility::Hidden;
```
We should do the content visibility check first, since it’s usually not hidden.

Why does `autoRevealsWhenFound` need to be precomputed and stored on the RenderStyle? Why not just check         `m_document->settings().detailsAutoExpandEnabled() && is<HTMLDetailsElement>(element)` here, through an appropriate helper function? Similarly, why do the `detailsAutoExpandEnabled()` check twice?

nt1m replied:
Same here for the bit. Though I can move the content-visibility check first.

darinadler commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -37,6 +37,7 @@
 #include "FontCascade.h"
 #include "HTMLAttachmentElement.h"
 #include "HTMLBodyElement.h"
+#include "HTMLDetailsElement.h"
```
I don't see what's used from this header in this file.

nt1m replied:
Good catch, it was a leftover from my older version of this PR where I made a DOM lookup instead of checking a style bit.

darinadler commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -455,7 +464,10 @@ static bool isRendererAccessible(const RenderObject* renderer, TextIteratorBehav
+        if (style.usedContentVisibility() == ContentVisibility::Auto)
+            return true;
+        if (renderer->document().settings().detailsAutoExpandEnabled() && renderer->style().autoRevealsWhenFound())
+            return true;
```
This code always returns true. Need to return false if neither of these conditions is true. This also makes me realize we don't have sufficient testing since no test failed.

nt1m replied:
Good catch! I'll take care of it here: https://bugs.webkit.org/show_bug.cgi?id=295035

darinadler commented on Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.h:
```diff
@@ -221,6 +221,8 @@ void postSynchronousMessage(const char* name, const void* value) = delete;
 void postPageMessage(const char* name);
 void postPageMessage(const char* name, bool value);
+void postPageMessage(const char* name, unsigned value);
+void postPageMessage(const char* name, uint64_t value);
```
Why mix `unsigned` and `uint64_t` here? Better to use `uint32_t` and `uint64_t` or `unsigned` and `unsigned long`.

nt1m replied:
This was just copy-paste from the code above. I'm not sure it puts both types. I'll just use `uint64_t`

darinadler replied:
It can be valuable to overload for more than one integer type, so that the width of the integer isn't changed by automatic type conversion. If we have an overload for more than one individual type then the comparable fail because it's ambiguous, which is desirable. However, I think we're OK with just the 64 bit one for now.

darinadler commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -218,4 +218,23 @@ void HTMLDetailsElement::toggleOpen()
     setBooleanAttribute(HTMLNames::openAttr, !hasAttributeWithoutSynchronization(HTMLNames::openAttr));
 }
 
+void expandAllDetailsAncestors(Node& node)
```
This function name sounds like it expands more than it does: it doesn't walk out of `summary` elements, presumably for good reason. Might be better to have a name that is more accurate, if it can still be short and clear. Even just taking out the word _all_ might be an improvement.

darinadler commented on Source/WebKit/WebProcess/WebPage/FindController.cpp:
```diff
@@ -504,7 +505,9 @@ void FindController::didFindString()
-    selectedFrame->checkedSelection()->revealSelection();
+    CheckedRef selection = selectedFrame->selection();
+    selection->revealSelection();
+    WebCore::expandAllDetailsAncestors(*selection->selection().start().protectedAnchorNode());
```
Does argument dependent lookup obviate the need for the `WebCore::` prefix here?

darinadler commented on Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
```diff
@@ -160,6 +161,8 @@ void WebFoundTextRangeController::decorateTextRangeWithStyle(const WebFoundTextR
         case FindDecorationStyle::Highlighted: {
             m_highlightedRange = range;
 
+            WebCore::expandAllDetailsAncestors(simpleRange->protectedStartContainer());
```
Does argument dependent lookup obviate the need for the `WebCore::` prefix here?

darinadler commented on Source/WebKit/WebProcess/WebPage/ios/FindControllerIOS.mm:
```diff
@@ -180,6 +181,7 @@ static void setSelectionChangeUpdatesEnabledInAllFrames(WebPage& page, bool enab
     // text, so we reveal the text at the center of the viewport.
     // FIXME: Find a better way to estimate the obscured area (https://webkit.org/b/183889).
     frame->selection().revealSelection(SelectionRevealMode::RevealUpToMainFrame, ScrollAlignment::alignCenterAlways, WebCore::RevealExtentOption::DoNotRevealExtent);
+    WebCore::expandAllDetailsAncestors(*frame->selection().selection().start().anchorNode());
```
Does argument dependent lookup obviate the need for the `WebCore::` prefix here?

darinadler commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -455,7 +464,10 @@ static bool isRendererAccessible(const RenderObject* renderer, TextIteratorBehav
-        return style.usedContentVisibility() == ContentVisibility::Auto;
+        if (style.usedContentVisibility() == ContentVisibility::Auto)
+            return true;
+        if (renderer->document().settings().detailsAutoExpandEnabled() && renderer->style().autoRevealsWhenFound())
```
The code that sets the `RenderStyle` bit already checks the settings, so can we leave out the settings check here?

nt1m replied:
Yeah true, it's not needed here.

darinadler commented on Source/WebCore/editing/TextIterator.cpp:
```diff
@@ -466,7 +478,14 @@ static bool isConsideredSkippedContent(const RenderBox* renderBox, TextIteratorB
+    if (behaviors.contains(TextIteratorBehavior::EntersSkippedContentRelevantToUser)) {
+        if (renderBox->style().usedContentVisibility() != ContentVisibility::Hidden)
+            return false;
+        if (renderBox->document().settings().detailsAutoExpandEnabled() && renderBox->style().autoRevealsWhenFound())
```
The code that sets the `RenderStyle` bit already checks the settings, so can we leave out the settings check here?

nt1m replied:
Yep I can remove it here too

darinadler commented on Source/WebCore/html/HTMLDetailsElement.cpp:
```diff
@@ -218,4 +219,26 @@ void HTMLDetailsElement::toggleOpen()
     setBooleanAttribute(HTMLNames::openAttr, !hasAttributeWithoutSynchronization(HTMLNames::openAttr));
 }
 
+// https://html.spec.whatwg.org/#ancestor-details-revealing-algorithm
```
The algorithm in the spec does the toggling during the iteration, not separately after building a list of details elements. Is that observably different in any unusual edge cases?

nt1m replied:
It doesn't make a difference. I was just being cautious since we usually don't mutate during iteration, but it's probably more relevant when you use the special iterators, which is not the case now that I switched to a while loop

nt1m replied:
Changed to match the spec

nt1m committed:
Auto-expand “details” element for find-in-page and scrolling to fragments
https://bugs.webkit.org/show_bug.cgi?id=228843
rdar://81856620

Reviewed by Darin Adler.

This change causes any closed <details> element to be opened
automatically (auto-expanded) if either:

- the user is using find-in-page and a descendant node of the closed
  <details> element contains a match for the given search string
- the user is navigating to a fragment ID and the fragment ID is for a
  descendant element of the closed <details> element

…per the requirements in the HTML spec’s Find-in-page section at
https://html.spec.whatwg.org/#interaction-with-details-and-hidden=until-found

Otherwise, without this change, when a descendant node of a closed
<details> element contains a match for the given find-in-page search
string, or when navigating to a fragment ID for a descendant element of
a closed <details> element, the closed <details> element isn’t opened
automatically (not auto-expanded).

Changes to TextIterator ensure that content inside details elements are
not skipped over for the purposes of find-in-page.

This change also introduces the DetailsAutoExpandEnabled preference
(disabled by default), for controlling whether <details> elements get
auto-expanded under the conditions described above.

And this also adds a testRunner.indicateMatchIndex(index) function —
to emulate jumping through matches in the non-findString codepath.

Original PR by: Michael[tm] Smith <mike@w3.org>

* LayoutTests/TestExpectations:
* LayoutTests/editing/find/cocoa/find-and-replace-in-closed-details-expected.txt: Added.
* LayoutTests/editing/find/cocoa/find-and-replace-in-closed-details.html: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-closed-details-expected.txt: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-closed-details.html: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-summary-of-closed-details-expected.txt: Added.
* LayoutTests/editing/text-iterator/find-in-page-in-summary-of-closed-details.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-details-element/auto-expand-details-element-fragment-expected.txt:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/editing/TextIterator.cpp:
(WebCore::TextIterator::TextIterator):
(WebCore::isRendererAccessible):
(WebCore::isConsideredSkippedContent):
* Source/WebCore/editing/TextIteratorBehavior.h:
* Source/WebCore/html/HTMLDetailsElement.cpp:
(WebCore::revealClosedDetailsAncestors):
* Source/WebCore/html/HTMLDetailsElement.h:
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::scrollToFragmentInternal):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::autoRevealsWhenFound const):
* Source/WebCore/rendering/style/RenderStyleSetters.h:
(WebCore::RenderStyle::setAutoRevealsWhenFound):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
(WebCore::StyleRareInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/style/StyleAdjuster.cpp:
(WebCore::Style::Adjuster::adjust const):
* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(WKPageIndicateFindMatch):
* Source/WebKit/UIProcess/API/C/WKPage.h:
* Source/WebKit/WebProcess/WebPage/FindController.cpp:
(WebKit::FindController::didFindString):
* Source/WebKit/WebProcess/WebPage/WebFoundTextRangeController.cpp:
(WebKit::WebFoundTextRangeController::decorateTextRangeWithStyle):
* Source/WebKit/WebProcess/WebPage/ios/FindControllerIOS.mm:
(WebKit::FindController::didFindString):
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::postPageMessage):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.h:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::indicateFindMatch):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveMessageFromInjectedBundle):

Canonical link: https://commits.webkit.org/296708@main
